src/state-visualizer/visualizer.py

```diff
@@ -0,0 +1,126 @@
+from __future__ import annotations
+import graphviz
+import time
+from PyQt5.QtWidgets import *
+from PyQt5.QtCore import *
+from PyQt5.QtGui import QPainter
+from PyQt5.QtSvg import QSvgRenderer
+
+import rospy
+from smach_msgs.msg import SmachContainerStatus, SmachContainerStructure
+from threading import Lock
+from dataclasses import dataclass
+from typing import Optional, List, Dict
+
+
+@dataclass
+class State:
+    name: str
+    children: List[State]
+
+
+class StateMachine:
+    def __init__(self):
+        self.states: Dict[str, State] = {}
+        self.structure: Optional[SmachContainerStructure] = None
+        self.mutex: Lock = Lock()
+        self.cur_active: str = ""
+        self.needs_redraw: bool = True
+
+    def set_active_state(self, active_state):
+        """
+        sets the state specified to be active (thread safe)
+        """
+        with self.mutex:
+            if active_state != self.cur_active and active_state in self.states:
+                self.cur_active = active_state
+                self.needs_redraw = True
+
+    def __rebuild(self, structure: SmachContainerStructure):
+        """
+        rebuilds the state dictionary with a new structure message
+        """
+        self.states = {child: State(child, []) for child in structure.children}
+        for start, end in zip(structure.outcomes_from, structure.outcomes_to):
+            if end != "None":
+                self.states[start].children.append(self.states[end])
+        self.needs_redraw = True
+
+    def check_rebuild(self, structure: SmachContainerStructure):
+        """
+        checks if the structure passed as input matches the structure already represented (thread safe)
+        """
+        with self.mutex:
+            if structure == self.structure:
+                return False
+            else:
+                self.__rebuild(structure)
+                self.structure = structure
+
+    def container_status_callback(self, status: SmachContainerStatus):
+        self.set_active_state(status.active_states[0])
+
+    def container_structure_callback(self, structure: SmachContainerStructure):
+        self.check_rebuild(structure)
+
+
+class GUI(QWidget):  # type: ignore
+    def __init__(self, state_machine):
```

Comment:
this isn't a problem.

Reply:
https://stackoverflow.com/questions/20125172/what-is-the-problem-with-shadowing-names-defined-in-outer-scopes/20125739#20125739

It is idiomatic in Rust for example but in Python weird things can happen with scoping (since dynamic scoping is used)

I don't care too much but pycharm warns against this for a valid reason. It also makes it harder to rename

Reply:
i get it but I just think its stupid. Having to come up with different (and ultimately less-semantic) names for the same objects for every level of indirection/scope reduces readability. This "rule" is not followed in most python code that I've seen.

Reply:
yeah that is the issue I typically run into as well, finding a good name. stupid python scoping

Reply:
Tell PyCharm to fuck off then and lets merge this bad boy

Reply:
![now-reality-can-be-whatever-i-want](https://user-images.githubusercontent.com/10037572/188710708-49cd1d95-b385-44e1-8a0d-1c490457c819.gif)

Reply:
the beauty of python really